webapp/advantage/views.py

```diff
@@ -747,6 +764,117 @@ def account_view(**kwargs):
+    marketplace = kwargs.get("marketplace")
+
+    advantage = UAContractsAPI(
+        session, token, api_url=api_url, is_for_view=True
+    )
+
+    accounts = advantage.get_accounts(email=email)
+
+    total_payments = []
+    for account in accounts:
+        raw_payments = advantage.get_account_purchases(
+            account_id=account["id"],
+            filters={"marketplace": marketplace},
+        )
+
+        if not raw_payments:
+            continue
+
+        product_listings = raw_payments["productListings"]
+        for raw_payment in raw_payments["purchases"]:
+            payment_id = raw_payment["id"]
+            payment_marketplace = raw_payment["marketplace"]
+            created_at = raw_payment["createdAt"]
+            total = raw_payment["invoice"]["total"] / 100
+            currency = raw_payment["invoice"]["currency"]
+            listing_id = raw_payment["purchaseItems"][0]["productListingID"]
+
+            period = None
+            for product_listing in product_listings:
+                if product_listing["id"] == listing_id:
+                    period = product_listing["period"]
+
+                    break
+
+            file_name = _get_download_file_name(raw_payment, period)
+            download_link = f"invoices/download/{period}/{payment_id}"
+
+            total_payments.append(
+                {
+                    "created_at": created_at,
+                    "service": services[payment_marketplace]["name"],
+                    "period": "Monthly" if period == "monthly" else "Annual",
+                    "date": parse(created_at).strftime("%d %B, %Y"),
+                    "total": f"{total} {currency}",
+                    "download_file_name": file_name,
+                    "download_link": download_link,
+                }
+            )
+
+    total_payments.sort(key=lambda item: item["created_at"], reverse=True)
+
+    return flask.render_template(
+        "account/invoices/index.html",
+        invoices=total_payments,
+        marketplace=marketplace,
+        is_test_backend=is_test_backend,
+    )
+
+
+@advantage_checks(check_list=["is_maintenance", "view_need_user"])
+def download_invoice(period, purchase_id, **kwargs):
+    token = kwargs.get("token")
+    api_url = kwargs.get("api_url")
+
+    advantage = UAContractsAPI(
+        session, token, api_url=api_url, is_for_view=True
+    )
+
+    purchase = advantage.get_purchase(purchase_id)
+    file_name = _get_download_file_name(purchase, period)
+    download_link = purchase["invoice"]["url"]
+
+    response = urllib.request.urlopen(download_link)
+
+    return flask.Response(
+        response.read(),
+        mimetype="application/pdf",
+        headers={"Content-Disposition": f"attachment;filename={file_name}"},
+    )
+
+
+def _get_download_file_name(payment, period):
+    payment_marketplace = payment["marketplace"]
+    created_at = payment["createdAt"]
+
+    file_name_elements = [services[payment_marketplace]["short"]]
```

Comment:
Where does services come from? I don't see it in the function definition

Reply:
I extracted it as a global variable at the beginning of `views.py`.
 
 https://github.com/canonical-web-and-design/ubuntu.com/pull/10062/files/60a15b39f47be74145a91c65efd7299d6d1d04ee#diff-aa9ecfa0f8c508e3003b04b0ffe83ba3698ed5f0a70237517932822a9a1a09e9R39-R52
 
I thought I should have a global place to define all the marketplaces to short name, full name conversions so they are consistent. I am not sure it's the best/nicest way to do it though.

Reply:
This is fair enough, but in this case you would normally want to signify that it's [a constant](https://www.python.org/dev/peps/pep-0008/#constants), with capital letters.

`SERVICES[payment_marketplace]["short"]`

To make it clear that it shouldn't be manipulated within any functions.

Reply:
Fixed!